proposals/0000-linear-types.rst

```diff
@@ -63,125 +69,45 @@ Linear types enable better solutions to both problems:
-2. that after freeing the associated pointer can never be read.
-
-With these two guarantees in hand, users no longer need to rely on the
-GC for managing all resources, hence benefiting from lower tail
-latencies and potentially higher throughput, while still getting
-freedom from segfaults.
-
-Linear types don't just enable using Haskell for more use cases
-(low-latency trading appliances, low-level services in
-high-performance scientific computing clusters, etc). Correctly
-tracking the lifecycle of I/O resources has been a vexing issue for
-many network services. Creating a variant of the BSD socket API that
-statically guarantees ordering constraints between API calls becomes
-possible without the overhead of heavyweight encodings based *e.g.* on
-parameterized monads (see the `paper
-<https://arxiv.org/abs/1710.09756>`_ for more on this example).
-
-::
-
-  -- We need an variant of the IO monad where actions are linear
-  data RIO a
-  returnL :: a ->. RIO a
-  bindL :: RIO a ->. (a ->. RIO b) ->. RIO b
-
-  -- Definition of sockets
-  data State = Unbound | Bound | Listening | Connected
-  data Socket (s :: State)
-  data SocketAddress
-
-  -- When a (TCP) socket is created it is Unbound.
-  socket :: RIO (Socket Unbound)
-  -- To bind a socket to a port we take an Unbound socket, and make it
-  -- Bound. The type of bindL will ensure that the socket is threaded
-  -- through the computation, so that the (Socket Unbound) is not
-  -- accessible: we cannot bind a socket twice.
-  bind :: Socket Unbound ->. SocketAddress -> RIO (Socket Bound)
-  -- A socket must be bound to a port before we start listening
-  listen :: Socket Bound->. RIO (Socket Listening)
-  -- A socket can accept multiple connection, therefore, the socket is
-  -- returned in the same state by accept. A second, bidirectional,
-  -- socket representing the connection is also returned. Both have to
-  -- be used in a single-threaded fashion.
-  accept :: Socket Listening ->. RIO (Socket Listening, Socket Connected)
-  connect :: Socket Unbound ->. SocketAddress -> RIO (Socket Connected)
-  send :: Socket Connected ->. ByteString -> RIO (Socket Connected, Unrestricted Int)
-  receive :: Socket Connected -> RIO (Socket Connected, Unrestricted ByteString)
-  close :: ∀s. Socket s -> RIO ()
+In the `companion paper <https://arxiv.org/abs/1710.09756>`_ to this
+proposal, we have worked out in detail several use cases for linear
+types. We argue that linear types have far ranging consequences for
+the language. Salient use cases from the paper include:
+
+* Safe mutable arrays with a safe *non-copying* `freeze` operation.
+* Off-heap memory that enables allocating, reading, writing and
+  freeing memory safely, without use-after-free or double-free errors.
+  This is an important use case for latency sensitive systems
+  programming, where moving objects off-heap, out of the purview of
+  the GC, is beneficial for avoiding long GC pauses and achieving
+  predictable latencies.
```

Comment:
I don't think we have this in the paper (I made it part of the Haskell Exchange talk, though).